File: Algorithms/1005.Maximize-Sum-Of-Array-After-K-Negations/maximize_sum_of_array_after_k_negations.py
class Solution:
    def largestSumAfterKNegations(self, A: List[int], K: int) -> int:
        if K == 0:
            return sum(A)
        if len(A) == 0:
            return 0
        
        neg_num = sum([1 if ele < 0 else 0 for ele in A])
        if neg_num > K:
            A.sort()
            for index, num in enumerate(A):
                if index < K:
                    A[index] = -A[index]
                else:
                    break
            return sum(A)
        else:
            flag = (K - neg_num) % 2
            A = list(map(abs, A))
            if flag:
                return sum(A) - min(A)*2
            else:
                return sum(A)
        

File: Algorithms/1005.Maximize-Sum-Of-Array-After-K-Negations/maximize_sum_of_array_after_k_negations.go
func largestSumAfterKNegations(A []int, K int) int {
    if K == 0 {
        return 0
    }
    min := 101
    sort.Ints(A)
    index := 0
    for K > 0 {
        if index >= len(A) {
            break
        }
        if A[index] >= 0 {
            if A[index] < min {
                min = A[index]
            }
            break
        }
        if A[index] < 0 {
            A[index] = -A[index]
        }
        if A[index] < min {
            min = A[index]
        }
        index++
        K--
    }
    
    total := 0
    for _, val := range A {
        total += val
    }
    if K % 2 == 0 {
        return total
    } else {
        return total - 2*min
    }
    
}